src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -70,3 +111,192 @@ def dss_deletion_subscription_query(self, prefix: str) -> JSON:
+                raise RequirementError(f'Unexpected link_type: {link_type}')
+
+        def retrieve_rows(entity_type: str, entity_ids: Set[str]):
+            pk_column = entity_type + '_id'
+            non_pk_columns = bundle.data_columns if entity_type.endswith('_file') else bundle.metadata_columns
+            columns = ', '.join(non_pk_columns | {pk_column})
+            uuid_in_list = ' OR '.join(f'{pk_column} = "{entity_id}"' for entity_id in entity_ids)
+            rows = self._query_latest_version(f'''
+                SELECT {columns}
+                FROM {self._source.bq_name}.{entity_type}
+                WHERE {uuid_in_list}
+            ''', group_by=pk_column)
+            log.info('Retrieved %s %s entities', len(rows), entity_type)
+            return rows
+
+        with ThreadPoolExecutor(max_workers=config.num_repo_workers) as executor:
+            futures = {
+                entity_type: executor.submit(retrieve_rows, entity_type, entity_ids)
+                for entity_type, entity_ids in entities.items()
+            }
+        for entity_type, future in futures.items():
+            rows = future.result()
+            entity_ids = entities[entity_type]
+            for i, row in enumerate(rows):
+                bundle.add_entity(f'{entity_type}_{i}.json', entity_type, row)
+                entity_ids.remove(row[entity_type + '_id'])
+            if entity_ids:
+                raise RuntimeError(f'Required entities not found in {self._source.bq_name}.{entity_type}: {entity_ids}')
+        return bundle
+
+
+class Checksums(NamedTuple):
+    crc32c: str
+    sha1: str
+    sha256: str
+    s3_etag: str
+
+    def asdict(self) -> Dict[str, str]:
+        return self._asdict()
+
+    @classmethod
+    def without_values(cls) -> Dict[str, str]:
+        return {f: None for f in cls._fields}
+
+    @classmethod
+    def extract(cls, json: JSON) -> 'Checksums':
+        return cls(**{f: json[f] for f in cls._fields})
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class TDRBundle(Bundle):
+
+    def add_entity(self, entity_key: str, entity_type: str, entity_row: BigQueryRow) -> None:
+        content_type = 'links' if entity_type == 'links' else entity_row['content_type']
+        self.manifest.append(self._manifest_entry(name=entity_key,
+                                                  uuid=entity_row[entity_type + '_id'],
+                                                  version=entity_row['version'].strftime(Plugin.timestamp_format),
+                                                  size=entity_row['content_size'],
+                                                  content_type='application/json',
+                                                  dcp_type=f'"metadata/{content_type}"'))
+        if entity_type.endswith('_file'):
+            descriptor = json.loads(entity_row['descriptor'])
+            drs_path = None if entity_row['drs_template'] is None else entity_row['drs_template'].rsplit('/', 1)[1]
```

Comment:
Please use a URL parser and assert the fake domain name. Since their way of using a fake DNS is so stupid I want us to break if any of that changes.

Reply:
In more recent snapshots the bogus domain name is replaced with the actual domain name (jade.datarepo-dev.broadinstitute.org). Should I require that this matches the TDR service URL instead?

Reply:
(This means that `file_id` column now contains complete valid DRS URIs, but I'm guessing we're still going to only index the path part for flexibility with the DSS plugin).